File: concepts/04_collections_and_iteration/06_generators_laziness_and_early_termination/r/test_02_closing_and_delegation.R
# polyglot-family: collections_and_iteration
# polyglot-concept: generators_laziness_and_early_termination
# polyglot-related: languages/r/language/11_conditions_control_flow_and_cleanup/
# polyglot-related+: test_086_on_exit_multiple_cleanup_actions.R
#
# 共同问题：提前终止时生产者如何关闭资源，嵌套生成怎样委托。
# 对照观察：R 没有 generator close/delegation protocol；资源生产函数用 `on.exit`，组合用普通函数调用。

trace <- character()
produce <- function() {
    on.exit(trace <<- c(trace, "closed"), add = TRUE)
    c(1L, 2L)
}
delegate <- function() c(produce(), 3L)
result <- delegate()

stopifnot(
    identical(result, 1:3),
    identical(trace, "closed")
)

File: concepts/04_collections_and_iteration/06_generators_laziness_and_early_termination/ruby/test_02_closing_and_delegation.rb
# 共同问题：提前停止时谁拥有生产者清理，嵌套迭代怎样委托。
# 输入：带 ensure 的 each、break、外部 Enumerator 和 flat_map；观察：栈内清理确定性及外部状态边界。
# polyglot-family: collections_and_iteration
# polyglot-concept: generators_laziness_and_early_termination
# polyglot-related: languages/ruby/language/06_collections_and_enumeration/
# polyglot-related+: test_045_each_enumerable_and_external_enumerators.rb

require "assertions"

A = PolyglotAssertions

trace = []
iterable = Object.new
iterable.define_singleton_method(:each) do |&consumer|
  return enum_for(__method__) unless consumer

  begin
    consumer.call(1)
    consumer.call(2)
  ensure
    trace << :closed
  end
end

A.equal(1, iterable.each { |value| break value })
A.equal([:closed], trace)

trace.clear
external = iterable.each
A.falsey(external.respond_to?(:close))
A.equal(1, external.next)
A.equal(2, external.next)
A.raises(StopIteration) { external.next }
A.equal([:closed], trace)

delegated = [[1, 2], [3]].lazy.flat_map(&:itself)
A.equal([1, 2, 3], delegated.force)

A.done
